File: vasp_poscar_analyzer.f90
program vasp_poscar_analyzer
    implicit none

    ! Variable declarations
    integer :: i
    double precision, parameter :: PI = 3.141592653589793238
    double precision :: factor, a_v(3), b_v(3), c_v(3)
    double precision :: a, b, c, alpha, beta, gamma, volume, rad_to_deg
    character(len=256) :: poscar

    ! Conversion factor from radians to degrees
    rad_to_deg = 180.0 / PI

    ! Retrieve POSCAR filename from command line argument
    call getarg(1, poscar)

    ! Read lattice vectors from POSCAR file
    open(unit=501, file=poscar, status="old")
    read(501,*)           ! Skip comment line
    read(501,*) factor    ! Scaling factor
    read(501,*) a_v       ! Lattice vector a
    read(501,*) b_v       ! Lattice vector b
    read(501,*) c_v       ! Lattice vector c
    close(501)

    ! Scale lattice vectors by the factor
    do i = 1, 3
        a_v(i) = factor * a_v(i)
        b_v(i) = factor * b_v(i)
        c_v(i) = factor * c_v(i)
    end do

    ! Calculate magnitudes of lattice vectors
    a = sqrt(sum(a_v**2))
    b = sqrt(sum(b_v**2))
    c = sqrt(sum(c_v**2))

    ! Compute cell volume using the triple product |a_v x b_v . c_v|
    volume = dot_product(cross_product(a_v, b_v), c_v)

    ! Calculate angles between lattice vectors in degrees
    alpha = acos(dot_product(b_v, c_v) / (b * c)) * rad_to_deg
    beta = acos(dot_product(c_v, a_v) / (c * a)) * rad_to_deg
    gamma = acos(dot_product(a_v, b_v) / (a * b)) * rad_to_deg

    ! Print results to screen
    print 100, 'a = ', a, '(Å)'
    print 100, 'b = ', b, '(Å)'
    print 100, 'c = ', c, '(Å)'
    print 100, 'alpha = ', alpha, '(°)'
    print 100, 'beta = ', beta, '(°)'
    print 100, 'gamma = ', gamma, '(°)'
    print 100, 'volume = ', volume, '(Å³)'

    ! Comma-separated values for easy export
    print 200, a, ',', b, ',', c, ',', alpha, ',', beta, ',', gamma, ',', volume

100 format (A9, F18.11, A15)
200 format (F12.4, A, F12.4, A, F12.4, A, F12.4, A, F12.4, A, F12.4, A, F12.4)

contains

    ! Function to compute cross product of two vectors
    function cross_product(a, b)
        double precision, dimension(3) :: cross_product
        double precision, dimension(3), intent(in) :: a, b

        cross_product(1) = a(2) * b(3) - a(3) * b(2)
        cross_product(2) = a(3) * b(1) - a(1) * b(3)
        cross_product(3) = a(1) * b(2) - a(2) * b(1)
    end function cross_product

end program vasp_poscar_analyzer

File: vasp_poscar_analyzer.py
import sys
import numpy as np

# Get the filename from the command line, defaulting to 'POSCAR'
file = sys.argv[1] if len(sys.argv) > 1 else 'POSCAR'

# Constants
rad_to_deg = 180 / np.pi

# Read the file once and close it, minimizing file I/O
with open(file, 'r') as target:
    data = target.read().splitlines()

# Scale factor and coordinate parsing
ax = float(data[1].strip())
R_v = np.array([list(map(float, data[i].split())) for i in range(2, 5)], dtype='float64') * ax

# Calculate lattice constants
a, b, c = np.linalg.norm(R_v, axis=1)

# Volume calculation using determinant for improved readability
volume = np.linalg.det(R_v)

# Dot products for angle calculations
axb = np.dot(R_v[0], R_v[1])
bxc = np.dot(R_v[1], R_v[2])
cxa = np.dot(R_v[2], R_v[0])

# Angle calculations
alpha = np.arccos(bxc / (b * c)) * rad_to_deg
beta = np.arccos(cxa / (c * a)) * rad_to_deg
gamma = np.arccos(axb / (a * b)) * rad_to_deg

# Output results with unified formatting
print(f"a = {a:18.11f} Å; {a * 0.1:18.11f} nm")
print(f"b = {b:18.11f} Å; {b * 0.1:18.11f} nm")
print(f"c = {c:18.11f} Å; {c * 0.1:18.11f} nm")
print(f"a = {a:.3f}; b = {b:.3f}; c = {c:.3f}\n")
print(f"a/c = {a/c:18.11f}")
print(f"c/a = {c/a:18.11f}\n")
print(f"{a:.4f}\t{b:.4f}\t{c:.4f}\t{a/c:.4f}\t{volume:.4f}\n")
print(f"alpha = {alpha:18.11f}")
print(f"beta  = {beta:18.11f}")
print(f"gamma = {gamma:18.11f}\n")
print(f"volume = {volume:18.11f} Å^3")
